pkg/ipamutil/reconciler.go

```diff
@@ -0,0 +1,228 @@
+	GetPool() client.Object
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ClaimReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
+	b := ctrl.NewControllerManagedBy(mgr).
+		WithEventFilter(predicates.ResourceNotPausedAndHasFilterLabel(ctrl.LoggerFrom(ctx), r.WatchFilterValue))
+
+	if err := r.Provider.SetupWithManager(ctx, b); err != nil {
+		return err
+	}
+	return b.Complete(r)
+}
+
+// Reconcile is called by the controller to reconcile a claim.
+func (r *ClaimReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Reconciling claim")
+
+	// Fetch the IPAddressClaim
+	claim := &ipamv1.IPAddressClaim{}
+	if err := r.Get(ctx, req.NamespacedName, claim); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if _, ok := claim.GetLabels()[clusterv1.ClusterNameLabel]; ok {
+		cluster, err := clusterutil.GetClusterFromMetadata(ctx, r.Client, claim.ObjectMeta)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				log.Info("IPAddressClaim linked to a cluster that is not found, unable to determine cluster's paused state, skipping reconciliation")
+				return ctrl.Result{}, nil
+			}
+
+			log.Error(err, "error fetching cluster linked to IPAddressClaim")
+			return ctrl.Result{}, err
+		}
+
+		if annotations.IsPaused(cluster, cluster) {
+			log.Info("IPAddressClaim linked to a cluster that is paused, skipping reconciliation")
+			return ctrl.Result{}, nil
+		}
+	}
+
+	// Create a patch helper for the claim.
+	patchHelper, err := patch.NewHelper(claim, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		if err := patchHelper.Patch(ctx, claim); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	// Ensure the claim has a finalizer.
+	controllerutil.AddFinalizer(claim, ReleaseAddressFinalizer)
+
+	// Create the provider handler and fetch the pool.
+	handler := r.Provider.ClaimHandlerFor(r.Client, claim)
+	if res, err := handler.FetchPool(ctx); err != nil || res != nil {
+		if apierrors.IsNotFound(err) {
+			err := errors.New("pool not found")
+			log.Error(err, "the referenced pool could not be found")
+			if !claim.ObjectMeta.DeletionTimestamp.IsZero() {
+				return r.reconcileDelete(ctx, claim)
+			}
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, errors.Wrap(err, "failed to fetch pool")
+	}
+
+	if pool := handler.GetPool(); pool != nil && annotations.HasPaused(pool) {
+		log.Info("IPAddressClaim references Pool which is paused, skipping reconciliation.", "IPAddressClaim", claim.GetName(), "Pool", pool.GetName())
+		return ctrl.Result{}, nil
+	}
+
+	address := ipamv1.IPAddress{}
+	if err := r.Client.Get(ctx, req.NamespacedName, &address); err != nil && !apierrors.IsNotFound(err) {
+		return ctrl.Result{}, errors.Wrap(err, "failed to fetch address")
+	}
```

Comment:
I think we don't need this anymore, the allocation of `address` can be moved down to line 150. The `CreateOrPatch` method will fetch the address for us, and `reconcileDelete` fetches it as well.

Reply:
Done